### ISSUE TITLE
DEX auth token header length issues on v6.3.0

### ISSUE THREAD

**Issue by lrstanley** [concourse/concourse#5851]
## Summary

Just upgraded our dev environment to v6.3.0 from v5.7.0, and 2 users are reporting problems authenticating via the web interface, due to Nginx (sitting in front of Concourse) throwing:

```
400 Bad Request

Request Header Or Cookie Too Large
```

And related Nginx logs:

```
2020/07/06 15:32:09 [info] 26472#26472: *897589 client sent too long header line: "Cookie: _gcl_au=1.1.1376805454.1588963645; _ga=GA1.2.778376655.1588963645; DEV_concourse-dt=ffffffgg79cc1e0445525d0f4f58155e4e5a4a4t378b; DEC=%7B%22SEGMENT%22%3A%22RES%22%2C%22JU%22%3A%22TRUNCATED%22%2C%22STOREBANNEREXPDATE%22%3A%2208%2F07%2F2020%2010%3A57%3A27%22%7D; TLTSID=74645591441348700406164839997852; _A_source=source=localhost:4200|campaign=(not%20set)|medium=referral|content=/; _A_time=64.1588963644472.1593545391369; _uetvid=fa33028-d34-4c55-50f0-268c13ccdf0c; skymarshal_auth0="bearer eyJhbGciOiJSUzI<TRUNCATED>..." while reading client request headers, client: 1.2.3.4, server: , request: "GET /api/v1/resources HTTP/1.1", host: "ourcluster.domain.com", referrer: "https://ourcluster.domain.com/"
```

Most users aren't reporting problems -- this used to be a large problem pre-5.7.0 where the amount of teams we were in, caused very large headers, but this isn't the case (5 teams vs I can login with over 140 teams). I believe this may be related to the AD groups being in the headers. The users in mention have over 700 AD groups (715 and 721).

What I am not sure of, if this is simply due to other headers being applied, or if the headers from Concourse are truly too large. I also am not sure how I can read the auth header.

Is there any way that I can open the header to see what's taking up the most space? (not familiar with jwt format is being used, or if salts are being used, from db, etc). I don't know if AD groups are embedded in the header... Is there any way I can ensure only the ad groups that map to a team are included (i.e. they'd have to relogin to get new access, but that's better than not being able to login at all).

We have temporarily made an Nginx adjustment (from an otherwise 90% stock Nginx config), to double the header length. I don't know if this is advised or not, but it seems to be an intermediary fix that works. The adjustment:

```nginx
large_client_header_buffers 8 16k; # defaults to 4 and 8k
```

## Steps to reproduce

1. Deploy v6.3.0
2. Setup team with ldap access via direct user (non-ad groups)
3. Have user return member of 700 AD groups (could be less, not very sure), via AD.
4. Attempt to login via the UI.

## Expected results

Login successful. Possibly stronger splitting of auth headers (I see only one skymarshal_auth[0-9], maybe it should be splitting it even further?).

## Triaging info

* Concourse version: v6.3.0
* Browser (if applicable): Chrome (latest), IE (latest), Edge (almost latest).
* Did this used to work? Yes.

**Comment by gaigaslab-operations** [comment]
This appears to be introduced in 6.1.0.  I have tested on 6.0.0 and it works fine; 6.1.0, 6.2.0, and 6.3.0 all fail with the same 502-bad gateway.  We are using cf_auth on PCF/PAS/TAS.

**Comment by vito** [comment]
Thanks for the report. This is definitely introduced by v6.1.0, which switched the tokens from our own minted ones (containing the list of authorized teams) to the Dex minted ones (containing the list of all groups you're a member of via the connector).

> Is there any way that I can open the header to see what's taking up the most space? (not familiar with jwt format is being used, or if salts are being used, from db, etc). I don't know if AD groups are embedded in the header... Is there any way I can ensure only the ad groups that map to a team are included (i.e. they'd have to relogin to get new access, but that's better than not being able to login at all).

The tokens are JWT format, containing two base64-encoded segments joined by a `.`. If you decode the second segment you should see something like this:

```json
{"iss":"https://.../sky/issuer","sub":"...","aud":"concourse-web","exp":1593791596,"iat":1593705196,"at_hash":"...","email":"...","email_verified":true,"groups":["aliter","aliter:Owners","aliter:owners",...]}
```

The `groups` key is the big part.

There's no way to reduce the list down to only the relevant entries. That's how the tokens worked before, but folks complained about having to log out and back in for newly configured teams, and here we are, haha.

I think to fix this we'll probably have to store session information in the database somehow, so we can provide a much shorter token to the client. 🤔 I'll put this in the backlog and we'll try to figure something out.

**Comment by lrstanley** [comment]
And just to confirm, decoding it... it's massive, and it is in fact the `groups` field, for us.

**Comment by agurney** [comment]
Is there any hope of using the LDAP group-search flags to narrow down which groups are included? I don't mean listing out all the Concourse team ones, but maybe there are some aspects of your AD structure whereby you can eliminate a bunch of groups that you "know" won't be used for team auth config.

https://concourse-ci.org/ldap-auth.html#configuring-ldap-group-search

**Comment by aoldershaw** [comment]
Was able to repro the 502 bad gateway on HAProxy (from #5832) using the following `docker-compose` override:

```yaml
version: '3'

services:
  web:
    environment:
      CONCOURSE_EXTERNAL_URL: http://localhost
      CONCOURSE_GITHUB_CLIENT_ID: clientid
      CONCOURSE_GITHUB_CLIENT_SECRET: clientsecret

  haproxy:
    image: haproxy
    depends_on:
    - web
    ports:
    - "80:80"
    volumes:
    - ./haproxy.cfg:/usr/local/etc/haproxy/haproxy.cfg
```

and the following haproxy config:

```
global
    log /dev/log local0
    log localhost local1 notice
    maxconn 2000
    daemon
    # this took some playing around with to find the sweet spot where normal requests pass, but authenticated ones fail
    tune.bufsize 2800
 
defaults
    log global
    mode http
    option httplog
    option dontlognull
    retries 3
    timeout connect 5000
    timeout client 50000
    timeout server 50000
 
frontend http-in
    bind *:80
    default_backend webservers
 
backend webservers
    stats enable
    stats auth admin:admin
    stats uri /haproxy?stats
    balance roundrobin
    option httpchk
    option forwardfor
    option http-server-close
    server web1 web:8080 check
```

I'm able to login with local user just fine, but when I log in with github user, I get the 502 bad gateway on `/sky/callback`

Modifying the `/sky/callback` handler to *not* send back the ID token as the access token causes the request to pass (although no API calls work anymore).

I'm sure this also applies to the nginx issue - I just tested it with haproxy because the root cause of the error seemed less obvious

**Comment by lrstanley** [comment]
@agurney, that's possible, though with a company of over 45000 employees, and over 500 from widely different portions of the company logging into Concourse, trimming down may prove to be difficult (but again -- possible). We have only a handful which actually utilize AD groups to login, since most of our automation handles it at a user level.

If this proves to be a difficult problem to solve, we may have to go that route.

**Comment by lrstanley** [comment]
Another issue that another user seems to be experiencing... still v6.3.0

1. Cleared browser cache
2. Logged in via the UI
3. Login initiated in the CLI (on a remote host)
4. Copied token from UI to CLI
5. This reports "could not find a valid token"

After looking at the token, it's exactly 4096 characters (may just be coincidence, but I don't think so). Trying to decode it myself, it successfully decodes the JSON object that is base64 encoded, HOWEVER it also throws an error saying there isn't valid base64 at the end of the token (mine seems to do this too, but I can login via fly CLI properly). Almost like there is jumbled characters at the end, or the token is getting truncated in some way.

The odd thing about this user is that they aren't in that many groups, only **78**, and their token isn't anywhere near as long as any of the other users that were previously having problems (decoding it, there is nothing special, it's like everyone else's). I'm in 64 groups, and my token length is ~3500.

He's using the new Edge browser (chromium based), but has also tried the latest Chrome release.

The work @aoldershaw is doing may solve this, but going to put this info out there anyway.

**Comment by vito** [comment]
@lrstanley That specific issue was fixed in [v6.4.0](https://github.com/concourse/concourse/releases/tag/v6.4.0#5770) - it was a client-side issue with reading a string that long. (Turns out hoisting around large string values is a nontrivial task no matter where it's going.)

**Comment by lrstanley** [comment]
@vito -- thanks!
